File: apps/mobile/android/app/src/main/kotlin/com/securitypulse/widget/DailyCardModel.kt
// DailyCardModel.kt
// Security Pulse — Android widget data model
//
// Non-sensitive data model shared between the Flutter app and the Android widget.
// The Flutter app writes this to SharedPreferences.
// The Jetpack Glance widget reads it.
//
// IMPORTANT: Only non-sensitive fields are permitted.
// Never add: auth tokens, is_correct, explanation, PII, incident data.

package com.securitypulse.widget

import android.content.Context
import android.content.SharedPreferences
import org.json.JSONObject
import java.time.Instant

/**
 * Completion state of the daily scenario.
 */
enum class CompletionState {
    AVAILABLE,
    COMPLETED,
    NO_ASSIGNMENT,
    OFFLINE,
    SIGNED_OUT;

    companion object {
        fun fromString(value: String): CompletionState = when (value) {
            "available" -> AVAILABLE
            "completed" -> COMPLETED
            "no_assignment" -> NO_ASSIGNMENT
            "offline" -> OFFLINE
            "signed_out" -> SIGNED_OUT
            else -> OFFLINE
        }
    }
}

/**
 * Minimal model written to SharedPreferences by the Flutter app.
 * Read by the Jetpack Glance AppWidget.
 */
data class DailyCardModel(
    val scenarioId: String?,
    val title: String,
    val completionState: CompletionState,
    val deepLinkRoute: String?,
    val expiresAt: Instant?,
) {
    companion object {
        private const val PREFS_NAME = "security_pulse_widget"
        private const val KEY_DAILY_CARD = "daily_card_model"
        private const val DEEP_LINK_SCHEME = "securitypulse"

        /**
         * Reads the current model from SharedPreferences.
         * Returns a signed-out fallback if no data is present.
         */
        fun read(context: Context): DailyCardModel {
            val prefs: SharedPreferences = context.getSharedPreferences(PREFS_NAME, Context.MODE_PRIVATE)
            val json = prefs.getString(KEY_DAILY_CARD, null) ?: return signedOut()

            return try {
                val obj = JSONObject(json)
                val expiresAtStr = obj.optString("expiresAt", "")
                val expiresAt = if (expiresAtStr.isNotEmpty()) Instant.parse(expiresAtStr) else null

                // Check expiry
                if (expiresAt != null && Instant.now().isAfter(expiresAt)) {
                    return DailyCardModel(
                        scenarioId = obj.optString("scenarioId").takeIf { it.isNotEmpty() },
                        title = obj.optString("title", "Security Pulse"),
                        completionState = CompletionState.OFFLINE,
                        deepLinkRoute = obj.optString("deepLinkRoute").takeIf { it.isNotEmpty() },
                        expiresAt = expiresAt,
                    )
                }

                DailyCardModel(
                    scenarioId = obj.optString("scenarioId").takeIf { it.isNotEmpty() },
                    title = obj.optString("title", "Security Pulse"),
                    completionState = CompletionState.fromString(
                        obj.optString("completionState", "signed_out")
                    ),
                    deepLinkRoute = obj.optString("deepLinkRoute").takeIf { it.isNotEmpty() },
                    expiresAt = expiresAt,
                )
            } catch (e: Exception) {
                signedOut()
            }
        }

        private fun signedOut() = DailyCardModel(
            scenarioId = null,
            title = "Sign in to see today's question",
            completionState = CompletionState.SIGNED_OUT,
            deepLinkRoute = null,
            expiresAt = null,
        )
    }

    /** Returns the deep link URI string for this card, or null if unavailable. */
    val deepLinkUri: String?
        get() = deepLinkRoute?.let { "$DEEP_LINK_SCHEME://$it" }
}

File: apps/mobile/ios/SecurityPulseWidget/DailyCardModel.swift
//
//  DailyCardModel.swift
//  SecurityPulseWidget
//
//  Non-sensitive data model shared between the Flutter app and the iOS widget.
//  The Flutter app writes this to the App Group UserDefaults.
//  The widget extension reads it via TimelineProvider.
//
//  IMPORTANT: This struct must only contain non-sensitive fields.
//  Never add: auth tokens, is_correct, explanation, employee PII,
//  incident data, or any confidential corporate content.
//

import Foundation

/// Completion state of the daily scenario.
enum DailyCardCompletionState: String, Codable {
    case available
    case completed
    case noAssignment = "no_assignment"
    case offline
    case signedOut = "signed_out"
}

/// Minimal model written to App Group storage by the Flutter app.
/// Read by the WidgetKit TimelineProvider.
struct DailyCardModel: Codable {
    /// UUID of the scenario. Used to construct the deep link.
    let scenarioId: String?
    /// Short title of the scenario (max 80 characters).
    let title: String
    /// Current completion state.
    let completionState: DailyCardCompletionState
    /// Deep link route to open in the Flutter app.
    let deepLinkRoute: String?
    /// When this cache entry expires. Widget shows offline state after this.
    let expiresAt: Date?

    /// Returns the deep link URL for this card, or nil if unavailable.
    var deepLinkURL: URL? {
        guard let route = deepLinkRoute else { return nil }
        return URL(string: "securitypulse://\(route)")
    }

    /// App Group identifier — must match the entitlement in both targets.
    /// Replace with your organisation's App Group identifier.
    static let appGroupIdentifier = "group.com.securitypulse.shared"

    /// UserDefaults key for the daily card model.
    static let defaultsKey = "daily_card_model"

    /// Reads the current model from App Group UserDefaults.
    /// Returns a signed-out fallback if no data is present.
    static func read() -> DailyCardModel {
        guard
            let defaults = UserDefaults(suiteName: appGroupIdentifier),
            let data = defaults.data(forKey: defaultsKey),
            let model = try? JSONDecoder().decode(DailyCardModel.self, from: data)
        else {
            return DailyCardModel(
                scenarioId: nil,
                title: "Sign in to see today's question",
                completionState: .signedOut,
                deepLinkRoute: nil,
                expiresAt: nil
            )
        }

        // Check expiry
        if let expiresAt = model.expiresAt, expiresAt < Date() {
            return DailyCardModel(
                scenarioId: model.scenarioId,
                title: model.title,
                completionState: .offline,
                deepLinkRoute: model.deepLinkRoute,
                expiresAt: model.expiresAt
            )
        }

        return model
    }
}
